src/mbgl/style/expression/in.cpp

```diff
@@ -0,0 +1,116 @@
+#include <string.h>
+#include <mbgl/style/conversion_impl.hpp>
+#include <mbgl/style/expression/in.hpp>
+#include <mbgl/style/expression/type.hpp>
+#include <mbgl/util/string.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+EvaluationResult In::evaluate(const EvaluationContext& params) const {
+    const EvaluationResult evaluatedNeedle = needle->evaluate(params);
+    const EvaluationResult evaluatedHeystack = haystack->evaluate(params);
+
+    if (!evaluatedNeedle) {
+        return evaluatedNeedle.error();
+    }
+    if (!evaluatedHeystack) {
+        return evaluatedHeystack.error();
+    }
+
+    type::Type evaluatedNeedleType = typeOf(*evaluatedNeedle);
+    if (!isComparableRuntimeValue(evaluatedNeedleType)) {
+        return EvaluationError{"Expected first argument to be of type boolean, string or number, but found " +
+                               toString(evaluatedNeedleType) + " instead."};
+    }
+
+    type::Type evaluatedHeystackType = typeOf(*evaluatedHeystack);
+    if (!isSearchableRuntimeValue(evaluatedHeystackType)) {
+        return EvaluationError{"Expected second argument to be of type array or string, but found " +
+                               toString(evaluatedHeystackType) + " instead."};
+    }
+
+    if (evaluatedNeedleType == type::Null || evaluatedHeystackType == type::Null) {
+        return EvaluationResult(false);
+    }
+
+    if (evaluatedHeystackType == type::String) {
+        const auto heystackString = evaluatedHeystack->get<std::string>();
+        std::string needleValue = "";
+        if (evaluatedNeedleType == type::Boolean) {
+            needleValue = evaluatedNeedle->get<bool>() ? "true" : " false";
+        } else if (evaluatedNeedleType == type::String) {
+            needleValue = evaluatedNeedle->get<std::string>();
+        } else if (evaluatedNeedleType == type::Number) {
+            needleValue = std::to_string(evaluatedNeedle->get<double>());
+            needleValue.erase(needleValue.find_last_not_of('0') + 1, std::string::npos);
+            needleValue.erase(needleValue.find_last_not_of('.') + 1, std::string::npos);
+        }
+        return EvaluationResult(heystackString.find(needleValue) != std::string::npos);
+    } else {
+        const auto heystackArray = evaluatedHeystack->get<std::vector<Value>>();
```

Comment:
I think this will crash if `evaluatedHeystack` is `Null`. Could you add unit tests?

Reply:
We have checked Null in `isSearchableRuntimeValue` in line 29 to insure it will not happens here.

Reply:
What's more, we already have test case that Haystack is null like [assert-string](https://github.com/mapbox/mapbox-gl-js/blob/c82a0b8c89bb964824996e25f54c31cfaf419e61/test/integration/expression-tests/in/assert-string/test.json)

Reply:
> We have checked Null in isSearchableRuntimeValue in line 29 to insure it will not happens here.

Why isSearchableRuntimeValue returns true for Null type? And then, evaluation returns if it is Null type in https://github.com/mapbox/mapbox-gl-native/pull/16162/files#diff-5d51b639578982530d698f5b5fd0536eR49